File: old/104-max-depth-tree/js/104.js
#!/usr/bin/env node

const TreeNode = val => ({
    val:  val,
    left: null,
    right: null,
});

const maxDepth = root => {
    if (root == null) {
        return 0;
    }
    return 1 + Math.max( maxDepth(root.left), maxDepth(root.right) );
};

const t1 = TreeNode(3);
t1.left = TreeNode(9);
t1.right = TreeNode(20);
t1.right.left = TreeNode(15);
t1.right.right = TreeNode(7);

const expected_result = 3;
const actual_result = maxDepth(t1);
console.log("Expected result:", expected_result);
console.log("Actual result:", actual_result);
console.log("Equal?", expected_result == actual_result);

File: old/104-max-depth-tree/elixir/104.exs
#! /usr/bin/env elixir

defmodule BinTree do
  defstruct value: nil, left: nil, right: nil
  def node(v, l, r), do: %BinTree{value: v, left: l, right: r}
  def leaf(v), do: %BinTree{value: v, left: nil, right: nil}
end

defmodule Solution do
  def max_depth(nil), do: 0

  def max_depth(%BinTree{} = t) do
    1 + max(max_depth(t.left), max_depth(t.right))
  end
end

alias BinTree, as: BT
t1 = BT.node(3, BT.leaf(9), BT.node(20, BT.leaf(15), BT.leaf(7)))

IO.puts("Expected result: 3")
IO.puts("Actual result:")

t1
|> Solution.max_depth()
|> IO.inspect()
